types/lualogging/logging.d.tl

```diff
@@ -0,0 +1,110 @@
+local enum Level
+   "DEBUG"
+   "INFO"
+   "WARN"
+   "ERROR"
+   "FATAL"
+   "OFF"
+end
+
+local type Append = function<T>(params: T, ...: any): Log
+
+local record Log
+   append: Append
+   setLevel: function (self: Log, level: Level)
+   log: function(self: Log, level: Level, ...: any)
+   getPrint: function(self: Log, level: Level): function(...: any)
+
+   debug: function(self: Log, ...: any) 
+   info: function(self: Log, ...: any) 
+   warn: function(self: Log, ...: any) 
+   error: function(self: Log, ...: any) 
+   fatal: function(self: Log, ...: any) 
+
+   DEBUG: Level
+   INFO: Level
+   WARN: Level
+   ERROR: Level
+   FATAL: Level
+   OFF: Level
+end
+
+local enum ConsoleDestination
+   "stdout"
+   "stderr"
+end
+
+local record ConsoleParameters
+   destination: ConsoleDestination
+   logPattern: string
+   logPatterns: { Level: string }
+   timestampPattern: string
+   logLevel: Level
+end
+
+local record FileParameters
+   filename: string
+   datePattern: string
+   logPattern: string
+   logPatterns: { Level: string }
+   timestampPattern: string
+   logLevel: Level
+end
+
+local record RollingFileParameters
+   filename: string
+   maxFileSize: number
+   maxBackupIndex: number
+   logPattern: string
+   logPatterns: { Level: string }
+   timestampPattern: string
+   logLevel: Level
+end
+
+local record Types
+   Level: Level
+   Append: Append
+   Log: Log
+   ConsoleDestination: ConsoleDestination
+   ConsoleParameters: ConsoleParameters
+   FileParameters: FileParameters
+   RollingFileParameters: RollingFileParameters
+end
+
+local record logging
+   _COPYRIGHT: string
+   _DESCRIPTION: string
+   _VERSION: string
+
+   DEBUG: Level
+   INFO: Level
+   WARN: Level
+   ERROR: Level
+   FATAL: Level
+   OFF: Level
+
+   new: function(append: function(self: Log, level: Level, msg: string | function), startLevel: Level): Log
+   buildLogPatterns: function(patterns: table, default: string): table
+   defaultLogPatterns: function(patt: string | table): table
+   defaultTimestampPattern: function(patt: string): string
+   defaultLevel: function(level: Level): Level
+   defaultLogger: function(logger: Log): Log
+
+   -- undocumented/private APIs
+   compilePattern: function(pattern: any): function(): string
+   prepareLogMsg: function(lpattern: string, dpattern: string, level: Level, message: string): string
+   tostring: function(value: any): string
+
+   -- Deprecated
+   getDeprecatedParams: function(lst: table, ...: any)
+
+   -- Appenders (dynamically added to logging when required)
+   console: Append<ConsoleParameters>
+   file: Append<FileParameters>
+   rolling_file: Append<RollingFileParameters>
+   -- TODO : add more appenders
+
+   __types: Types
```

Comment:
I expose the types with this additional property. I think it's ok to add this property (that doesn't exist in the library) because it's "teal only", and doesn't add anything that could be translated to Lua.

In the usage, the user can now refer to types from this file somewhere else in their code: `local myParams: logging.__types.ConsoleParameters = { foo }`. The usage of `__types` will not be translated to anything in Lua, so there is no impact for adding this property.

Note that I named it `__types` with this double `_` to make it feels like a meta property. I have no strong opinion about it.

Reply:
I think you could just put those in the module's main record. In other definition files from this repo, we just went ahead and exposed `PascalCase` type names as part of the public API (see for example [luafilesystem](https://github.com/teal-language/teal-types/blob/master/types/luafilesystem/lfs.d.tl)), assuming those won't clash with the usual `lowercase` or `snake_case` entries exposed by the Lua APIs we're typing. For the Teal user, using `logging.ConsoleParameters` rather than `logging.__types.ConsoleParameters` would be a bit more comfortable, I think, and I believe clashes with future lualogging APIs would be extremely unlikely.

Reply:
Thank you for pointing out the luafilesystem example. I have checked some other types definition to see how similar needs were solved. I somehow missed it...

I actually like the idea to include the inner-record directly in the record. It gives some scope to it. The `PascalCase` should solve the clashes with properties defined by the module API (for a lot of Lua module), and help to identify the property as a type name (this “convention” is also used by typescript). So I'm all in!